File: backend/app/services/positions.py
"""Orden de lectura de las posiciones dentro de una subcaja (docs/FORMULARIO.md):
cartón se lee de arriba hacia abajo y de izquierda a derecha (columna `1..9`, fila
`A..I`); plástica se lee de izquierda a derecha y de arriba hacia abajo (`1..100`
en orden numérico)."""

from app.models import BoxType

_CARTON_ROWS = "ABCDEFGHI"


def position_order(box_type: str) -> list[str]:
    if box_type == BoxType.CARTON_81.value:
        return [f"{number}{letter}" for number in range(1, 10) for letter in _CARTON_ROWS]
    if box_type == BoxType.PLASTIC_100.value:
        return [str(number) for number in range(1, 101)]
    raise ValueError(f"Tipo de caja desconocido: {box_type}")


def next_free_position(box_type: str, occupied: set[str]) -> str | None:
    for position in position_order(box_type):
        if position not in occupied:
            return position
    return None

File: frontend/src/utils/positions.ts
import type { BoxType } from "../api/types";

// Espejo de app/services/positions.py: orden de lectura de las posiciones
// dentro de una subcaja (docs/FORMULARIO.md). Cartón se lee de arriba hacia
// abajo y de izquierda a derecha (columna 1..9, fila A..I); plástica se lee
// de izquierda a derecha y de arriba hacia abajo (1..100 en orden numérico).
const CARTON_ROWS = "ABCDEFGHI".split("");

export function positionOrder(boxType: BoxType): string[] {
  if (boxType === "carton_81") {
    const positions: string[] = [];
    for (let number = 1; number <= 9; number += 1) {
      for (const letter of CARTON_ROWS) positions.push(`${number}${letter}`);
    }
    return positions;
  }
  const positions: string[] = [];
  for (let number = 1; number <= 100; number += 1) positions.push(String(number));
  return positions;
}

export function nextFreePosition(boxType: BoxType, occupied: ReadonlySet<string>): string | null {
  for (const position of positionOrder(boxType)) {
    if (!occupied.has(position)) return position;
  }
  return null;
}

const BOX_NAME_RE = /^([A-Ha-h])\s*(\d{1,3})$/;

/** Parte "Nombre Caja" (p. ej. `A12`) en letra de rack + N° de caja. */
export function parseBoxName(raw: string): { rackLetter: string; boxNumber: number } | null {
  const match = BOX_NAME_RE.exec(raw.trim());
  if (!match) return null;
  const boxNumber = Number(match[2]);
  if (boxNumber <= 0) return null;
  return { rackLetter: match[1].toUpperCase(), boxNumber };
}
